provider/oracle/environ.go

```diff
@@ -0,0 +1,372 @@
+// the bootstrap process by transferring the tools and installing the
+// initial Juju controller.
+//
+// Bootstrap will use just one specific architecture because the oracle
+// cloud only supports amd64.
+func (o oracleEnviron) Bootstrap(
+	ctx environs.BootstrapContext,
+	params environs.BootstrapParams,
+) (*environs.BootstrapResult, error) {
+	// in order too make the entire bootstrap process prossible
+	// we must take into accounting some things:
+	// validate if we have a shape based on the bootstrap constraints
+	// and pick the right one
+	// validate if we have in the imagelist a image correspoding to the
+	// image tools specified.
+	// validate if we have already the ssh keys and if we don't have them
+	// upload them into the oracle infrstracture and enable the flag
+
+	logger.Infof("Loging into the oracle cloud infrastructure")
+	if err := o.p.client.Authenticate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// make api request to oracle cloud to give us a list of
+	// all supported shapes
+	shapes, err := o.p.client.AllShapeDetails()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// find the best suitable shape returned from the api
+	shape, err := findShape(shapes.Result, params.BootstrapConstraints)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	logger.Infof(
+		"Choosing the %s with %d cores and %d MB ram",
+		shape.name, shape.cpus, shape.ram,
+	)
+
+	// make api request to oracle cloud to give us a list of iamges that are
+	// stored there, if we don't have a image complaint with the metadata provided
+	// by juju, if the shape picked up above is supported, we should bail out
+	imagelist, err := checkImageList(o.p.client, params.ImageMetadata, shape)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// we are using just the juju ssh keys, no other keys
+	keys := strings.Split(o.cfg.AuthorizedKeys(), "\n")
+	// we should try to determine if we need to upload the keys or if
+	// there is already the set of keys there if we found the ssh key
+	// then we should check if the key has the flag enabled on true
+	// if not, change it to true
+	// if the key is not present in the oracle infrstracture then we should
+	// make a request and upload it to make use for further bootstraping the
+	// controller or adding other machines
+	nameKey, err := uploadSSHControllerKeys(o.p.client, keys[0])
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	instance, err := launchBootstrapConstroller(o.p.client, []oci.InstanceParams{
+		{
+			Shape:     shape.name,
+			Imagelist: imagelist,
+			Label:     o.cfg.UUID(),
+			SSHKeys:   []string{nameKey},
+			Name:      o.cfg.Name(),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_ = instance
+	return nil, nil
+}
+
+// BootstrapMessage optionally provides a message to be displayed to
+// the user at bootstrap time.
+func (o oracleEnviron) BootstrapMessage() string {
+	return "SomeBootstrapMessage"
```

Comment:
Just leave this blank. I'm going to delete that method, it never should have been introduced - we have a context param for bootstrap that can be used as needed